pkg/server/plugin/keymanager/azurekms/azurekms_test.go

```diff
@@ -0,0 +1,1048 @@
+			generatedKeyIds: []string{"key-1", "key-2", "key-3"},
+		},
+		{
+			name: "non existing keys",
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			// setup
+			ts := setupTest(t)
+
+			_, err := ts.plugin.Configure(ctx, configureRequestWithDefaults(t))
+			require.NoError(t, err)
+
+			// Generate the keys
+			for _, keyID := range tt.generatedKeyIds {
+				_, err = ts.plugin.GenerateKey(ctx, &keymanagerv1.GenerateKeyRequest{
+					KeyId:   keyID,
+					KeyType: keymanagerv1.KeyType_RSA_4096,
+				})
+				require.NoError(t, err)
+			}
+
+			// exercise
+			resp, err := ts.plugin.GetPublicKeys(ctx, &keymanagerv1.GetPublicKeysRequest{})
+
+			if tt.err != "" {
+				require.Error(t, err)
+				require.Equal(t, err.Error(), tt.err)
+				return
+			}
+
+			require.NotNil(t, resp)
+			require.NoError(t, err)
+			require.Equal(t, len(tt.generatedKeyIds), len(resp.PublicKeys))
+		})
+	}
+}
+
+func TestRefreshKeys(t *testing.T) {
+	entry1 := makeFakeKeyEntry(t, keyName+"-1", trustDomain, validServerID, azkeys.JSONWebKeyTypeRSA, nil, to.Ptr(4096))
+	entry2 := makeFakeKeyEntry(t, keyName+"-2", trustDomain, "another-server-id", azkeys.JSONWebKeyTypeRSA, nil, to.Ptr(4096))
+	entry3 := makeFakeKeyEntry(t, keyName+"-3", "another-td", validServerID, azkeys.JSONWebKeyTypeRSA, nil, to.Ptr(4096))
+	entry4 := makeFakeKeyEntry(t, keyName+"-4", "another-td", "another-server-id", azkeys.JSONWebKeyTypeRSA, nil, to.Ptr(4096))
+
+	for _, tt := range []struct {
+		name             string
+		configureRequest *configv1.ConfigureRequest
+		err              string
+		fakeEntries      []fakeKeyEntry
+		updateKeyErr     string
+	}{
+		{
+			name:             "refresh keys error",
+			configureRequest: configureRequestWithDefaults(t),
+			err:              "update failure",
+			updateKeyErr:     "update failure",
+			fakeEntries: []fakeKeyEntry{
+				makeFakeKeyEntry(t, keyName, trustDomain, validServerID, azkeys.JSONWebKeyTypeRSA, nil, to.Ptr(4096)),
+			},
+		},
+		{
+			name:             "refresh keys succeeds",
+			configureRequest: configureRequestWithDefaults(t),
+			fakeEntries: []fakeKeyEntry{
+				entry1,
+				entry2,
+				entry3,
+				entry4,
+			},
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			// setup
+			ts := setupTest(t)
+			ts.kmsClient.setEntries(tt.fakeEntries)
+			ts.kmsClient.setUpdateKeyErr(tt.updateKeyErr)
+			refreshKeysSignal := make(chan error)
+			ts.plugin.hooks.refreshKeysSignal = refreshKeysSignal
+
+			// exercise
+			_, err := ts.plugin.Configure(ctx, tt.configureRequest)
+			require.NoError(t, err)
+
+			// wait for refresh keys task to be initialized
+			_ = waitForSignal(t, refreshKeysSignal)
+			// move the clock forward so the task is run
+			ts.clockHook.Add(6 * time.Hour)
+			// wait for refresh keys to be run
+			err = waitForSignal(t, refreshKeysSignal)
+
+			// assert
+			if tt.updateKeyErr != "" {
+				require.NotNil(t, err)
+				require.Equal(t, tt.err, err.Error())
```

Comment:
you can use require.EqualError() that validate error is not nil

Reply:
These are strings we're comparing here though not errors